### PR TITLE
Application Linking

### PR DESCRIPTION
**Pull request by bafonins**
**Summary:**

This PR adds the Application Link page
Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/214

Not linked:
Screen Shot 2019-03-10 at 13 26 15

Frontend validation: 
Screen Shot 2019-03-10 at 13 29 27

Linking failed:
Screen Shot 2019-03-10 at 13 27 57

Linked:
Screen Shot 2019-03-10 at 13 27 23

**Changes:**

- Add the Application Link page
- Extend the `<Form />` component to keep the reset button enabled
- Add the `className` prop to the `<DataSheet />` component

### CODE REVIEW

**Comment by htdvisser** [comment]
I don't think "Leave empty if within a cluster" is clear enough, perhaps change to something like "Leave empty to link to the Network Server in the same cluster" (if this is more correct, @johanstokking)?

When exactly do we see that "not linked to `my-app`" error?

Typo: Downlink receives

**Comment by johanstokking** [comment]
Agreed with @htdvisser suggestion regarding cluster.

Regarding the error; you get the linking error itself from the stats endpoint. So if the API key is wrong, you'll get a permission denied there.

The problem there is however that you can't determine the difference between permission denied to call the stats rpc, or that the linking failed with permission denied. So we may want to return an "aborted" error with the linking failed error as message?

**Comment by htdvisser** [comment]
It would indeed make sense to return an `Aborted` error if the linking failed (for whatever reason)

**Comment by bafonins** [comment]
@johanstokking @htdvisser updated screenshots

**Comment by kschiffer** [comment]
I would propose to show the setup form and the link data next to each other. For this, we would need to detach the unlink button from the `<Form />` and make it separate, and as a consequence provide a way to reference the `<Formik />` component using a `formikRef` prop. We would need the reference to trigger `resetForm()` from outside the form context.
See my proposed changes [here](https://github.com/TheThingsNetwork/lorawan-stack/compare/feature/214-application-link...feature/214-application-link-changes).

image

File: pkg/webui/console/views/application-link/index.js [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+      await this.setState({
+        fetching: false,
+        error: '',
+        link,
+        stats,
+      })
+    } catch (error) {
+      // show only non-404 errors
+      if (error && error.code !== 5) {
```
[the remark this answers is not in the document]
Comment (bafonins) [reply pkg/webui/console/views/application-link/index.js]: why 500? 5 is the grpc `NotFound` error code, see https://github.com/grpc/grpc-go/blob/master/codes/codes.go

[the remark this answers is not in the document]
Comment (kschiffer) [reply pkg/webui/console/views/application-link/index.js]: Right, sorry, mixed that up with HTML status codes.

**Pull request merged**